File: python/tests/test_handshake.py
"""Transport: handshake, attach guarantee, replay-on-attach, reconnect."""

from __future__ import annotations

from typing import Any

from graphmind.protocol import PROTOCOL_VERSION

from .conftest import wait_until


def test_hello_carries_protocol_capabilities_and_sdk(attached: Any) -> None:
    _instance, viewer = attached()
    hello = viewer.wait_for_type("hello")

    assert hello["gm"] == PROTOCOL_VERSION
    assert hello["runId"] == "*"
    payload = hello["payload"]
    assert payload["versions"]["protocol"] == PROTOCOL_VERSION
    assert payload["versions"]["client"]
    assert set(payload["capabilities"]) == {"pause", "step", "inject", "retry", "abort"}
    assert payload["sdk"]["name"] == "python"


def test_ready_resolves_true_once_attached(attached: Any) -> None:
    instance, _viewer = attached()
    assert instance.attached is True
    # A second call short-circuits.
    assert instance.ready(timeout=0.1) is True


def test_ready_returns_false_when_no_viewer_listens(make_gm: Any) -> None:
    instance = make_gm(url="ws://127.0.0.1:1/ingest", connect_timeout=0.1)
    assert instance.ready(timeout=1.0) is False
    assert instance.attached is False


def test_ready_returns_false_when_disabled(make_gm: Any) -> None:
    instance = make_gm(url="ws://127.0.0.1:1/ingest", enabled=False)
    assert instance.ready(timeout=0.2) is False


async def test_ready_async_resolves(viewer: Any, make_gm: Any) -> None:
    view = viewer()
    instance = make_gm(url=view.url)
    assert await instance.ready_async(timeout=5.0) is True


def test_breakpoints_and_mode_are_armed_from_hello_ack(attached: Any) -> None:
    instance, _viewer = attached(breakpoints=[{"kind": "tool", "name": "search"}], mode="step")
    breakpoints, mode = instance.session._engine.snapshot()
    assert breakpoints == [{"kind": "tool", "name": "search"}]
    assert mode == "step"


def test_replay_on_attach_resends_buffered_events_with_original_seq(
    viewer: Any, make_gm: Any
) -> None:
    view = viewer()
    instance = make_gm(url="ws://127.0.0.1:1/ingest", connect_timeout=0.05)

    # Emitted while detached: buffered only.
    with instance.run("early"):
        pass
    assert instance.stats().buffered >= 4
    assert view.frames() == []

    # Now point a second instance at the live viewer and re-emit the buffer by
    # attaching: the client replays everything it still holds.
    instance.session._transport._url = view.url
    assert instance.ready(timeout=5.0) is True

    wait_until(lambda: len(view.of_type("run.started")) >= 1, label="replayed run.started")
    replayed = view.of_type("run.started")[0]
    assert replayed["payload"]["meta"]["name"] == "early"
    # Original seq preserved so viewers can deduplicate (decisions.md #5).
    seqs = [f["seq"] for f in view.frames() if f["type"] != "hello"]
    assert seqs == sorted(seqs)


def test_version_mismatch_keeps_the_client_detached(viewer: Any, make_gm: Any) -> None:
    view = viewer(ack_protocol=PROTOCOL_VERSION + 1)
    instance = make_gm(url=view.url)
    assert instance.ready(timeout=1.5) is False
    assert instance.attached is False


def test_handshake_timeout_keeps_the_client_detached(viewer: Any, make_gm: Any) -> None:
    view = viewer(auto_ack=False)
    instance = make_gm(url=view.url, handshake_timeout=0.2)
    assert instance.ready(timeout=1.5) is False
    assert instance.attached is False


def test_reconnect_after_viewer_restart(viewer: Any, make_gm: Any) -> None:
    view = viewer()
    instance = make_gm(url=view.url, retry_interval=0.2)
    assert instance.ready(timeout=5.0) is True

    view.kill_abruptly()
    wait_until(lambda: instance.attached is False, timeout=5.0, label="detach")

    # ready() kicks an immediate reconnect instead of waiting out the interval.
    replacement = viewer()
    instance.session._transport._url = replacement.url
    assert instance.ready(timeout=5.0) is True

File: ruby/test/test_handshake.rb
# frozen_string_literal: true

require_relative "test_helper"

class TestHandshake < Minitest::Test
  def test_connects_and_completes_the_handshake
    viewer = new_viewer
    session = new_session(url: viewer.url)

    assert session.ready(5.0), "expected ready() to report attached"
    assert session.attached?

    hello = viewer.wait_for_frame("hello").first
    assert_equal 1, hello["gm"]
    assert_equal "*", hello["runId"]
    assert_equal 1, hello["payload"]["versions"]["protocol"]
    assert_equal Graphmind::VERSION, hello["payload"]["versions"]["client"]
    assert_equal %w[pause step inject retry abort run-claim], hello["payload"]["capabilities"]
    # First connection: nothing to prove continuity with yet.
    refute hello["payload"].key?("resumeToken")
    assert_valid_frame(hello)
  end

  # The run-claim capability: the debugger mints a token in `hello.ack`, and a
  # reconnecting client echoes it so its runs are recognised as still its own.
  # Without it, a reconnect looks like a different process and the debugger
  # refuses to let it keep streaming the run it was already streaming.
  def test_echoes_the_session_token_on_reconnect
    viewer = new_viewer
    session = new_session(url: viewer.url, retry_interval: 0.1)
    assert session.ready(5.0)
    assert_nil viewer.wait_for_frame("hello").first["payload"]["resumeToken"]

    viewer.kill_abruptly
    wait_until(label: "detach") { !session.attached? }
    assert session.ready(5.0), "expected the client to reconnect"

    hellos = viewer.wait_for_frame("hello", count: 2)
    assert_operator hellos.length, :>=, 2, "expected a second hello after the reconnect"
    assert_equal FakeViewer::SESSION_TOKEN, hellos.last["payload"]["resumeToken"]
  end

  def test_ready_is_false_when_nothing_is_listening
    # Port 1 is reserved and nothing local answers it.
    session = new_session(url: "ws://127.0.0.1:1/ingest", connect_timeout: 0.3)

    refute session.ready(1.0)
    refute session.attached?
  end

  def test_ready_is_false_when_disabled
    session = new_session(url: "ws://127.0.0.1:1/ingest", enabled: false)

    refute session.ready(0.2)
    refute session.attached?
  end

  def test_adopts_breakpoints_and_mode_from_the_ack
    viewer = new_viewer(breakpoints: [{ "kind" => "tool", "name" => "search" }], mode: "step")
    session = new_session(url: viewer.url)
    assert session.ready(5.0)

    node = Graphmind::GateNode.new("tool:search", "tool", "search")
    other = Graphmind::GateNode.new("tool:book", "tool", "book")
    engine = session.instance_variable_get(:@engine)

    assert engine.should_pause?("before", node)
    # step mode pauses at every before point, matching or not
    assert engine.should_pause?("before", other)
    assert_equal "step", engine.mode
  end

  def test_refuses_a_viewer_on_a_different_protocol_version
    viewer = new_viewer(ack_protocol: 99)
    session = new_session(url: viewer.url)

    refute session.ready(1.5)
    refute session.attached?
    assert(warnings.any? { |w| w.include?("protocol") }, "expected a protocol warning, got #{warnings.inspect}")
  end

  def test_survives_a_server_that_refuses_the_upgrade
    viewer = new_viewer(refuse_upgrade: true)
    session = new_session(url: viewer.url)

    refute session.ready(1.5)
    refute session.attached?
  end

  def test_stays_detached_when_the_viewer_never_acks
    viewer = new_viewer(auto_ack: false)
    session = new_session(url: viewer.url, handshake_timeout: 0.3)

    refute session.ready(1.5)
    refute session.attached?
    assert(warnings.any? { |w| w.include?("handshake") }, warnings.inspect)
  end

  def test_reconnects_after_the_viewer_disappears_and_replays_the_buffer
    viewer = new_viewer
    session = new_session(url: viewer.url, retry_interval: 0.2)
    assert session.ready(5.0)

    session.emit("node.started", {
                   "nodeId" => "tool:a", "kind" => "tool", "name" => "a", "instanceId" => "i1"
                 })
    viewer.wait_for_frame("node.started")

    viewer.kill_abruptly
    wait_until(label: "detach") { !session.attached? }

    # Emitted while dark: must survive in the ring buffer.
    session.emit("node.finished", {
                   "nodeId" => "tool:a", "instanceId" => "i1", "durationMs" => 1.0, "status" => "ok"
                 })

    assert session.ready(5.0), "expected a reconnect"
    frames = viewer.wait_for_frame("node.finished")
    assert_equal 1, frames.length
    assert_operator viewer.connection_count, :>=, 2
    # Replayed envelopes keep their original seq so the viewer can dedupe.
    started = viewer.frames_of("node.started")
    assert_operator started.length, :>=, 1
  end
end
